core/src/main/java/org/jdbi/v3/CollectorFactoryRegistry.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jdbi.v3;
+
+import org.jdbi.v3.tweak.CollectorFactory;
+
+import java.util.Set;
+import java.util.SortedSet;
+import java.util.TreeSet;
+import java.util.concurrent.CopyOnWriteArraySet;
+import java.util.function.Function;
+import java.util.stream.Collector;
+
+/**
+ * Registry of collector factories.
+ * Contains a set of collector factories, registered by the application.
+ */
+class CollectorFactoryRegistry {
+
+    private final Set<CollectorFactory<?, ?, ?>> factories = new CopyOnWriteArraySet<>();
+
+    CollectorFactoryRegistry() {
+        factories.add(new SortedSetCollectorFactory());
+    }
+
+    CollectorFactoryRegistry(CollectorFactoryRegistry parent) {
+        factories.addAll(parent.factories);
+    }
+
+    CollectorFactoryRegistry createChild() {
+        return new CollectorFactoryRegistry(this);
+    }
+
+    void register(CollectorFactory<?, ?, ?> factory) {
+        factories.add(factory);
+    }
+
+    @SuppressWarnings("unchecked")
+    <T, R> Collector<T, ?, R> createCollectorFor(Class<R> type) {
+        for (CollectorFactory<?, ?, ?> factory : factories) {
+            if (factory.accepts(type)) {
+                return ((CollectorFactory<T, ?, R>) factory).newCollector(type);
+            }
+        }
+
+        throw new IllegalStateException("No collector builder available for " + type.getName());
+    }
+
+    private static class SortedSetCollectorFactory<T> implements CollectorFactory<T, SortedSet<T>, SortedSet<T>> {
+
+        public boolean accepts(Class<?> type) {
+            return type.equals(SortedSet.class);
```

Comment:
During refactoring I've realized that we don't need the accumulator type for container factories, because it's actually a detail of a collector implementation. 

The JDK `Collectors.toList` method returns `Collector<T, ?, List<T>>`